### DISCUSSION TITLE
mfs performance calculating

### DISCUSSION THREAD

**Discussion by rusticek** [moosefs/moosefs#520] [General]
Hi,
 
i'm thinking of how to properly calculate performance of mfs cluster. For example

solution1
- 2x MFSCHS, each 3x HDD
- goal 2

solution2
- 3x MFSCHS, each 3x HDD
- goal 3

solution3
- 5x MFSCHS, each 8xHDD
- goal 2

Data are stored in 64MiB chunks. In an ideal environment if i will read 64MiB file which is stored exactly in that one specific chunk file, what IOPS do i get?
Solution1 - i will get IOPS equals to 2 HDDs? (file is stored on 2 MFCHS, so it will be read from two HDDs)
Solution2 - i will get IOPS equals to 3 HDDs? (file is stored on 3 MFCHS, so it it will be read from three HDDs)
Solution3 - i will get IOPS equals to 2 HDDs? (file is stored on 2 MFCHS, so it will be read still only from two HDDs because goal = 2) 

Does IOPS performance rely on goal set? Or am i thinking all wrong about this? Maybe 64MiB file would be always divided into smaller pieces across other chunks/HDDs? So IOPS will scale based on HDD quantity?

Thanks for explanation.

**Comment by chogata** [comment]
MooseFS always reads only from 1 copy of a chunk in a single read operation. Things that can change efficiency of reads are topology and preferred labels - basically 2 ways to ensure that the copy a particular client (mount) is reading is the closest (topologically in your network) copy to it, meaning shortest delays (latency) in reading. To read from 2 different copies at the same time, you need 2 read operations (2 processes or threads reading the same chunk in a file, either from one client/mount or from 2 different clients).
For reading a single chunk it doesn't matter how many chunk servers there are in your instance, only how many copies of one chunk are available. That is, one chunk in 3 copies will be read equally fast if there are 3 physical chunk servers or if there are 5 physical chunk servers. That's obvious.
The number of chunk servers is important when you start to perform many parallel operations on MooseFS - this is what MooseFS was designed for. Then, the more servers the better, because a) different chunks will have copies on different servers, so if one (or more) process needs to read chunk A and another process (or processes) needs to read chunk B, then the more servers there are, the lower is the probability that all copies of A are on the same servers as all copies of B, allowing the processes involved to read from different servers; b) if you perform write operations, especially of new data (creating new chunks), the more servers are available, the less operations are performed on one at one time.
Number of disks may increase performance of one machine, provided the rest of the hardware can "keep up" (main data bus, network card). But in general, in MooseFS you get the most result when you perform your i/o operations in parallel, not when you use cutting edge hardware but still perform sequential i/o operations in a single thread.
